pythran/optimizations/pattern_transform.py

```diff
@@ -66,27 +66,37 @@
                ast.Num(n=-1)],
          keywords=[])),
 
-    # X * X => X ** 2
-    (ast.BinOp(left=Placeholder(0), op=ast.Mult(), right=Placeholder(0)),
-     lambda: ast.BinOp(left=Placeholder(0), op=ast.Pow(), right=ast.Num(n=2))),
-
-    # a + "..." + b => "...".join((a, b))
-    (ast.BinOp(left=ast.BinOp(left=Placeholder(0),
-                              op=ast.Add(),
-                              right=ast.Str(Placeholder(1))),
-               op=ast.Add(),
-               right=Placeholder(2)),
-     lambda: ast.Call(func=ast.Attribute(
-         ast.Attribute(
-             ast.Name('__builtin__', ast.Load(), None),
-             'str',
-             ast.Load()),
-         'join', ast.Load()),
-         args=[ast.Str(Placeholder(1)),
-               ast.Tuple([Placeholder(0), Placeholder(2)], ast.Load())],
-         keywords=[])),
 ]
 
+# Dictionary with ast operator name as keys for each a list of tuple of
+#   (left, right, replacement) is defined.
+# replacement have to be a lambda function to have a new ast to replace when
+# replacement is inserted in main ast
+know_pattern_BinOp = {
+    ast.Mult.__name__ : [
+        (Placeholder(0), ast.Num(1), lambda: Placeholder(0)), # X * 1 => X
```

Comment:
Unfortunately, I relaized that even this simple transformation is not always correct:

```
x = [1,2]
y = x * 1
y[0] = 2
print(x, y)
```

Reply:
Arg, damn python.. I need to find a way to make it safe...
For my test case I need somehow the replacements 
```
y += x * 0 => y += 0 => pass
```